### PR TITLE
Predicate system

### PR DESCRIPTION
**Pull request by blaesus**
This PR creates a `Predicate` system in place of the current `Filter` system, in preparation for connecting with JavaScript interface (https://github.com/immux/immux/issues/37).

The enhancements over the current system are:-

## 1. Strongly-typed representation of conditions

Previously, a `Filter` is the combination of a list of operands and a list of operators. Their is no guarantee that the numbers match and the expression makes sense: the operand list may have 5 members while the operator list has 2 members. something like
`a==1 && b==2 || c==3 d==4 e==5`

Now the predicate type system make it impossible to express illegal constructions like the example above.

## 2. Deep field path

Previously, one can only compare the first-level data (this.field). Now multiple layers can be compared (this.field.subfield).

## 3. Binary serialization

Previously, `Filter` is serialized by converting it to a string and read the bytes. This approach is not efficient. For example, to express a combination of three conditions, `A&&B&&C`, we have to duplicate `&` symbol for four times. Now, `Predicate` is serialized with a normal binary method, and only one byte is necessary to link the conditions.

# Implementation

## 1. String parsing

Parsing a string representation (like "this.age >= 10") is done in two phase:
1. tokenize, which groups characters (like "this" or ">=") together,
2. parsing, which takes the tokens and combine them in a tree.

### CODE REVIEW

**Review by poppindouble — APPROVED** [review]
looks good to me, only minor modification required.

File: src/storage/executor/predicate.rs
Comment (blaesus) [reply src/storage/executor/predicate.rs:644]: Good point. Revised.

**Comment by blaesus** [comment]
@poppindouble I think this PR clashes with some of yours. I'll merge this one only after you merged yours as they came earlier.

**Comment by poppindouble** [comment]
> @poppindouble I think this PR clashes with some of yours. I'll merge this one only after you merged yours as they came earlier.

yes you are right, thanks. will finish my stuff ASAP

**Pull request merged**
